File: hw90/src/main/java/com/example/hw90/MainActivity.java
package com.example.hw90;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;


public class MainActivity extends AppCompatActivity  {
    public static final String CONTACT_ID = "CONTACT ID";
    public static final String ITEM = "ITEM";
    public static final int ADD_PHONE_CONTACT = 1;
    public static final int ADD_EMAIL_CONTACT = 2;
    public static final int EDIT_CONTACT = 3;

    private ContactViewModel contactViewModel;

    private FloatingActionButton add;
    private FloatingActionButton addPhoneNumber;
    private FloatingActionButton addEmail;
    private boolean clicked = false;

    private EditText search;
    private List <Contact> mItems = new ArrayList<>();
    private ContactListAdapter mAdapter = new ContactListAdapter();
    private RecyclerView recyclerView;

    private Contact contact;
    private int contactId;
    private Contact contactForPhoneUpdate;
    private Contact contactForEmailUpdate;

    public String q = "Tom";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        add = findViewById(R.id.add_contact_button);
        addPhoneNumber = findViewById(R.id.add_phone_button);
        addEmail = findViewById(R.id.add_email_button);
        search = findViewById(R.id.search);

        recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setAdapter(mAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        contactViewModel = new ViewModelProvider(this).get(ContactViewModel.class);
        contactViewModel.getFilteredContact(q);
        contactViewModel.getAllContacts().observe(this, new Observer<List<Contact>>() {
            @Override
            public void onChanged(List<Contact> contacts) {
                mAdapter.setContacts(contacts);
            }
        });

        add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                updateVisibility();
            }
        });

        addPhoneNumber.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startAddPhoneContact();
                updateVisibility();
            }
        });

        addEmail.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startAddEmailContact();
                updateVisibility();
            }
        });

        mAdapter.setMListener(new ContactListAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(Contact contact) {
                Intent intent = new Intent(MainActivity.this, EditContact.class);
                intent.putExtra(ITEM, contact);
                startActivityForResult(intent, EDIT_CONTACT);
            }
        });

        //        Search start ----------------------------------------------------------------------------------
        search.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }
            @Override
            public void afterTextChanged(Editable s) {
                q = s.toString();
                contactViewModel.getFilteredContact(q);
                mAdapter.notifyDataSetChanged();
            }
        });
    }

    private void updateVisibility() {
        clicker(clicked);
        clicked = !clicked;
    }

    private void clicker(boolean clicked) {
        if (!clicked) {
            addPhoneNumber.setVisibility(View.VISIBLE);
            addEmail.setVisibility(View.VISIBLE);
        } else {
            addPhoneNumber.setVisibility(View.INVISIBLE);
            addEmail.setVisibility(View.INVISIBLE);
        }
    }

    private void startAddPhoneContact() {
        Intent intent = new Intent(this, AddPhoneContact.class);
        startActivityForResult(intent, ADD_PHONE_CONTACT);
    }

    private void startAddEmailContact() {
        Intent intent = new Intent(this, AddEmailContact.class);
        startActivityForResult(intent, ADD_EMAIL_CONTACT);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == ADD_PHONE_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            String addedName = data.getStringExtra(AddPhoneContact.ADD_NAME);
            String addedPhoneNumber = data.getStringExtra(AddPhoneContact.ADD_PHONE_NUMBER);
           contactViewModel.insert(new Contact(R.drawable.phone_icon, addedName, addedPhoneNumber));

        } else if (requestCode == ADD_EMAIL_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            String addedName = data.getStringExtra(AddEmailContact.ADD_NAME_E);
            String addedEmail = data.getStringExtra(AddEmailContact.ADD_EMAIL);
            contactViewModel.insert(new Contact(R.drawable.email_icon, addedName, addedEmail));
        }

        if (requestCode == EDIT_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            contact = (Contact) data.getSerializableExtra(EditContact.ITEM_FOR_UPDATE_OR_REMOVE);
            contactId = data.getIntExtra(CONTACT_ID, 0);
            contactForPhoneUpdate = (Contact) data.getSerializableExtra(EditContact.UPDATED_PHONE_CONTACT);
            contactForEmailUpdate = (Contact) data.getSerializableExtra(EditContact.UPDATED_EMAIL_CONTACT);

            if (contact != null && contactForPhoneUpdate == null && contactForEmailUpdate == null) {
                contactViewModel.delete(contact);

            } else if (contact != null && contactForPhoneUpdate != null && contactForEmailUpdate == null) {
                contactForPhoneUpdate.setId(contactId);
                contactViewModel.update(contactForPhoneUpdate);

            } else if (contact != null && contactForPhoneUpdate == null && contactForEmailUpdate != null) {
                contactForEmailUpdate.setId(contactId);
                contactViewModel.update(contactForEmailUpdate);
            }
        }
    }
}

File: hw5/src/main/java/com/example/hw5/MainActivity.kt
package com.example.hw5

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {
    private var contactViewModel: ContactViewModel? = null
    private var clicked = false
    private val mAdapter: ContactListAdapter = ContactListAdapter()
    private var contact: Contact? = null
    private var contactId = 0
    private var contactForPhoneUpdate: Contact? = null
    private var contactForEmailUpdate: Contact? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        viewRecyclerView.adapter = mAdapter
        viewRecyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

        contactViewModel = ViewModelProvider(this).get<ContactViewModel>(ContactViewModel::class.java)
        contactViewModel!!.getAllContacts().observe(this, Observer<List<Contact>> { t -> mAdapter.setContacts(t) })

        viewAddContactButton.setOnClickListener { updateVisibility() }

        viewAddPhoneButton.setOnClickListener {
            startAddPhoneContact()
            updateVisibility()
        }

        viewAddEmailButton.setOnClickListener {
            startAddEmailContact()
            updateVisibility()
        }

        mAdapter.setMListener(object : ContactListAdapter.OnItemClickListener {
            override fun onItemClick(contact: Contact) {
                val intent = Intent(this@MainActivity, EditContact::class.java)
                intent.putExtra(ITEM, contact)
                startActivityForResult(intent, EDIT_CONTACT)
            }
        })
    }

    private fun updateVisibility() {
        clicker(clicked)
        clicked = !clicked
    }

    private fun clicker(clicked: Boolean) {
        if (!clicked) {
            viewAddPhoneButton.visibility = View.VISIBLE
            viewAddEmailButton.visibility = View.VISIBLE

        } else {
            viewAddPhoneButton.visibility = View.INVISIBLE
            viewAddEmailButton.visibility = View.INVISIBLE
        }
    }

    private fun startAddPhoneContact() {
        val intent = Intent(this, AddPhoneContact::class.java)
        startActivityForResult(intent, ADD_PHONE_CONTACT)
    }

    private fun startAddEmailContact() {
        val intent = Intent(this, AddEmailContact::class.java)
        startActivityForResult(intent, ADD_EMAIL_CONTACT)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == ADD_PHONE_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            val addedName = data.getStringExtra(AddPhoneContact.ADD_NAME)
            val addedPhoneNumber = data.getStringExtra(AddPhoneContact.ADD_PHONE_NUMBER)
            contactViewModel!!.insert(Contact(R.drawable.phone_icon, addedName!!, addedPhoneNumber!!))

        } else if (requestCode == ADD_EMAIL_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            val addedName = data.getStringExtra(AddEmailContact.ADD_NAME_E)
            val addedEmail = data.getStringExtra(AddEmailContact.ADD_EMAIL)
            contactViewModel!!.insert(Contact(R.drawable.email_icon, addedName!!, addedEmail!!))
        }

        if (requestCode == EDIT_CONTACT && resultCode == Activity.RESULT_OK && data != null) {
            contact = data.getSerializableExtra(EditContact.ITEM_FOR_UPDATE_OR_REMOVE) as Contact?
            contactId = data.getIntExtra(CONTACT_ID, 0)
            contactForPhoneUpdate = data.getSerializableExtra(EditContact.UPDATED_PHONE_CONTACT) as Contact?
            contactForEmailUpdate = data.getSerializableExtra(EditContact.UPDATED_EMAIL_CONTACT) as Contact?

            if (contact != null && contactForPhoneUpdate == null && contactForEmailUpdate == null) {
                contactViewModel!!.delete(contact!!)

            } else if (contact != null && contactForPhoneUpdate != null && contactForEmailUpdate == null) {
                contactForPhoneUpdate!!.id = contactId
                contactViewModel!!.update(contactForPhoneUpdate!!)

            } else if (contact != null && contactForPhoneUpdate == null && contactForEmailUpdate != null) {
                contactForEmailUpdate!!.id = contactId
                contactViewModel!!.update(contactForEmailUpdate!!)
            }
        }
    }

    companion object {
        const val CONTACT_ID = "CONTACT ID"
        const val ITEM = "ITEM"
        const val ADD_PHONE_CONTACT = 1
        const val ADD_EMAIL_CONTACT = 2
        const val EDIT_CONTACT = 3
    }
}
